include/circt/Scheduling/Problems.h

```diff
@@ -268,6 +268,82 @@ class CyclicProblem : public virtual Problem {
   virtual LogicalResult verify() override;
 };
 
+/// This class models the accumulation of physical propagation delays on
+/// combinational paths along SSA dependences within a scheduling problem's
+/// abstract time steps/cycles.
+///
+/// Each operator type is annotated with estimated values for incoming and
+/// outgoing delays. Combinational operators (zero-latency, no internal
+/// registers) have only a single propagation delay; this important special case
+/// is modeled by setting the incoming and outgoing delays to the same value.
+///
+/// A solution to this problem comprises physical start times for each operation
+/// that satisfy a given cycle time constraint.
+class ChainingProblem : public virtual Problem {
+private:
+  ProblemProperty<float> cycleTime;
+  OperatorTypeProperty<float> incomingDelay, outgoingDelay;
+  OperationProperty<float> physicalStartTime;
+
+public:
+  explicit ChainingProblem(Operation *containingOp) : Problem(containingOp) {}
+
+  /// The cycle time constrains the accumulated propagation delays along
+  /// combinational paths inside a time step.
+  Optional<float> getCycleTime() { return cycleTime; }
+  void setCycleTime(float time) { cycleTime = time; }
+
+  /// The incoming delay denotes the propagation time from the operand inputs to
+  /// either the result outputs (combinational operators) or the first internal
+  /// register stage.
+  Optional<float> getIncomingDelay(OperatorType opr) {
+    return incomingDelay.lookup(opr);
+  }
+  void setIncomingDelay(OperatorType opr, float delay) {
+    incomingDelay[opr] = delay;
+  }
+
+  /// The outgoing delay denotes the propagation time from either the operand
+  /// inputs (combinational operators) or the last internal register stage to
+  /// the result outputs.
+  Optional<float> getOutgoingDelay(OperatorType opr) {
+    return outgoingDelay.lookup(opr);
+  }
+  void setOutgoingDelay(OperatorType opr, float delay) {
+    outgoingDelay[opr] = delay;
+  }
+
+  /// The physical start time is computed by the scheduler. It is relative to
+  /// the beginning of the cycle that \p op starts in.
+  Optional<float> getPhysicalStartTime(Operation *op) {
```

Comment:
Maybe a better name for this that is more related to it's behavior?  getStartTimeInCycle() maybe?

Reply:
Done.